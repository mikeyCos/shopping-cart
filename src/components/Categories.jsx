import PropTypes from "prop-types";
import { useEffect, useState } from "react";
import { Link } from "react-router-dom";
import styles from "../styles/Categories.module.css";

const Categories = ({ categories }) => {
  return (
    <article>
      <ul className={styles["categories"]}>
        {categories.map((category) => {
          return (
            <li key={category}>
              <Link to={`category/${category.replaceAll(" ", "-")}`}>
                {category}
              </Link>
            </li>
          );
        })}
      </ul>
    </article>
  );
};

Categories.propTypes = {
  categories: PropTypes.array,
};

export default Categories;
