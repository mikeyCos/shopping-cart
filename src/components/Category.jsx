import {
  Link,
  useLocation,
  useOutletContext,
  useParams,
} from "react-router-dom";
import ProductCard from "./ProductCard";
import Modal from "./Modal";
import parseCategory from "../utilities/parseCategory";
import styles from "../styles/Category.module.css";

/* Renders products based on the useParams category value
 * useParams category value will fallback to 'all'
 */
const Category = () => {
  const { category = "all" } = useParams();
  const [products] = useOutletContext();
  const { pathname } = useLocation();
  const productKey = parseCategory(category);
  const headingCategory = category.replaceAll("-", " ");

  return (
    <section>
      <h2>Category</h2>
      <h3>{headingCategory}</h3>
      {products && (
        <>
          <section role="region" className={styles["products"]}>
            {products[productKey].map((product) => (
              <ProductCard key={product.id} className={styles["product-card"]}>
                <Link
                  to={`${pathname}/product/view/modal/${encodeURIComponent(
                    product.title
                  )}`}
                  state={{ product, previousLocation: pathname }}
                >
                  <ProductCard.Picture
                    src={product.image}
                    alt="#"
                    className={styles["product-picture"]}
                  />
                </Link>
                <ProductCard.Info
                  title={product.title}
                  price={product.price}
                  className={styles["product-card-info"]}
                />
              </ProductCard>
            ))}
          </section>
          <Modal />
        </>
      )}
    </section>
  );
};

export default Category;
