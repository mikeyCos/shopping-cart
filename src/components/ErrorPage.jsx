import { useRouteError } from "react-router-dom";

const ErrorPage = () => {
  const { error, status } = useRouteError();
  console.log(error);
  return (
    <section id="error-section" role="region">
      {error.message}
    </section>
  );
};

export default ErrorPage;
