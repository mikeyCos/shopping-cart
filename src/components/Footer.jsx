import "../styles/Footer.module.css";

const Footer = () => {
  return (
    <footer>
      <div>Placeholder</div>
    </footer>
  );
};

export default Footer;
