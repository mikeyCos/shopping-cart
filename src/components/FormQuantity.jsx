import PropTypes from "prop-types";
import styles from "../styles/FormQuantity.module.css";

const FormQuantity = ({ submitForm, children }) => {
  const onSubmitHandler = (e) => {
    e.preventDefault();
    submitForm();
  };

  return (
    <form
      noValidate={true}
      onSubmit={onSubmitHandler}
      className={styles["form-quantity"]}
    >
      {children}
    </form>
  );
};

const SubmitButton = ({ text }) => (
  <button type="submit" className={styles["submit-btn"]}>
    {text}
  </button>
);

FormQuantity.SubmitButton = SubmitButton;

FormQuantity.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.object,
    PropTypes.arrayOf(PropTypes.object),
  ]),
  submitForm: PropTypes.func.isRequired,
};

SubmitButton.propTypes = {
  text: PropTypes.string.isRequired,
};

export default FormQuantity;

/* 
// Product.jsx
<form noValidate={true} onSubmit={onSubmitHandler}>
  <p>{product.description}</p>
  <p>{formatPrice(product.price)}</p>
  <InputQuantity
    quantity={quantity}
    setQuantity={onChangeQuantityHandler}
    incrementHandler={incrementQuantityHandler}
    decrementHandler={decrementQuantityHandler}
  />
  <button type="submit">Add to cart</button>
</form>; */

/* 
// Cart.jsx
<form
  noValidate={true}
  onSubmit={(e) => {
    e.preventDefault();
    removeFromCart(item.id);
  }}
>
  <p>{formatPrice(item.price)}</p>
  <InputQuantity
    quantity={item.quantity}
    setQuantity={(newQuantity) => {
      console.log("previous quantity", item.quantity);
      setQuantity(item.id, newQuantity);
    }}
    incrementHandler={() => {
      const newQuantity = +item.quantity + 1;
      newQuantity <= 999 && setQuantity(item.id, newQuantity);
    }}
    decrementHandler={() => {
      const newQuantity = item.quantity - 1;
      newQuantity >= 0 && setQuantity(item.id, newQuantity);
    }}
  />
  <button type="submit">Delete</button>
</form>; */
