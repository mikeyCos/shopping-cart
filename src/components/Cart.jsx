import PropTypes from "prop-types";
import { createContext, useContext, useState } from "react";
import { Link } from "react-router-dom";
import ProductCard from "./ProductCard";
import FormQuantity from "./FormQuantity";
import InputQuantity from "./InputQuantity";
import formatPrice from "../utilities/formatPrice";
import styles from "../styles/Cart.module.css";

const Cart = () => {
  const { cart, removeFromCart, setQuantity } = useContext(CartContext);

  const subTotal = cart.reduce((accumulator, currentCartItem) => {
    return accumulator + currentCartItem.price * currentCartItem.quantity;
  }, 0);

  return (
    <section id={styles["cart"]} role="region">
      Cart
      {cart.length === 0 && <h4>Cart is empty</h4>}
      {cart.length > 0 && (
        <section role="region" className={styles["cart-container"]}>
          {cart.map((item) => {
            return (
              <ProductCard
                key={item.id}
                className={styles["cart-product-card"]}
              >
                <ProductCard.Heading title={item.title} />
                <Link
                  to={`/shop/product/view/${encodeURIComponent(item.title)}`}
                  state={{ product: item }}
                >
                  <ProductCard.Picture
                    src={item.image}
                    alt="#"
                    className={styles["product-picture"]}
                  />
                </Link>
                <FormQuantity submitForm={() => removeFromCart(item.id)}>
                  <p>{formatPrice(item.price)}</p>
                  <div className="form-controls">
                    <InputQuantity
                      quantity={item.quantity}
                      setQuantity={(newQuantity) => {
                        console.log("previous quantity", item.quantity);
                        setQuantity(item.id, newQuantity);
                      }}
                      incrementHandler={() => {
                        const newQuantity = +item.quantity + 1;
                        newQuantity <= 999 && setQuantity(item.id, newQuantity);
                      }}
                      decrementHandler={() => {
                        const newQuantity = item.quantity - 1;
                        newQuantity >= 0 && setQuantity(item.id, newQuantity);
                      }}
                    />
                    <FormQuantity.SubmitButton text="Delete" />
                  </div>
                </FormQuantity>
              </ProductCard>
            );
          })}
          <p className={styles["subtotal"]}>
            Subtotal: <span>{formatPrice(subTotal)}</span>
          </p>
        </section>
      )}
    </section>
  );
};

/* React hook createContext
 * https://react.dev/learn/passing-data-deeply-with-context
 */
const CartContext = createContext([]);

const CartProvider = ({ children }) => {
  const [cart, setCart] = useState([]);
  /* Removing item from cart
   * Check if item is in cart
   * If item is in cart
   *  Reduce cart and add quantities of item
   * Else
   *  Add item to cart
   *
   * Update item's quantity
   */
  const addToCart = (product, quantity) => {
    const { id } = product;

    /* // Option 1:
    const newCart = productInCart
      ? cart.map((productInCart) => {
          if (productInCart.id === id)
            return {
              ...productInCart,
              quantity: productInCart.quantity + quantity,
            };
          return productInCart;
        })
      : [...cart, { ...product, quantity }];
    setCart(newCart); */

    let isInCart = false;
    const newCart = cart.map((productInCart) => {
      if (productInCart.id === id) {
        isInCart = true;
        return {
          ...productInCart,
          quantity: productInCart.quantity + quantity,
        };
      }

      return productInCart;
    });

    setCart(isInCart ? newCart : [...cart, { ...product, quantity }]);

    /* // Option 2:
    setCart((prevCart) => {
      const isInCart = prevCart.some((cartItem) => cartItem.id === product.id);

      return isInCart
        ? prevCart.reduce((accumulator, currentCartItem) => {
            return [
              ...accumulator,
              currentCartItem.id === product.id
                ? {
                    ...currentCartItem,
                    quantity: currentCartItem.quantity + quantity,
                  }
                : currentCartItem,
            ];
          }, [])
        : [...prevCart, { ...product, quantity }];
    }); */
  };

  const removeFromCart = (productID) => {
    setTimeout(() => {
      setCart((prevCart) => {
        return prevCart.filter((product) => product.id !== productID);
      });
    }, 250);
  };

  const setQuantity = (productID, quantity) => {
    if (quantity == 0) {
      removeFromCart(productID);
    } else {
      setCart((prevCart) =>
        prevCart.map((product) => {
          return product.id === productID ? { ...product, quantity } : product;
        })
      );
    }
  };

  return (
    <CartContext.Provider
      value={{ cart, addToCart, removeFromCart, setQuantity }}
    >
      {children}
    </CartContext.Provider>
  );
};

CartProvider.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
};

export { Cart as default, CartContext, CartProvider };
