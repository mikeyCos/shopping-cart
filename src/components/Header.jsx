import { Link, NavLink } from "react-router-dom";
import { useContext } from "react";
import { CartContext } from "./Cart";
import NavAnchor from "./NavAnchor";
import CartIcon from "../assets/icons/cart-variant.svg?react";
import AppLogo from "../assets/illustrations/undraw_shopping_app.svg?react";
import GitHubIcon from "../assets/icons/github-mark/github-mark.svg?react";
import styles from "../styles/Header.module.css";
import iconStyles from "../styles/icons.module.css";
import illustrationsStyles from "../styles/illustrations.module.css";

const Header = () => {
  const { cart } = useContext(CartContext);
  return (
    <header className={styles["header"]}>
      <nav className={styles["nav"]}>
        <ul className={styles["nav-left"]}>
          <li>
            <Link to="/">
              <AppLogo className={illustrationsStyles["logo"]} />
              <h1>Shopping Cart App</h1>
            </Link>
          </li>
        </ul>

        <ul className={styles["nav-right"]}>
          <li>
            <NavAnchor pathname="home" />
          </li>

          <li>
            <NavAnchor pathname="shop" />
          </li>

          <li>
            <Link to="/cart" className={styles["cart-anchor"]}>
              <span className={styles["cart-number"]}>{cart.length}</span>

              <CartIcon
                className={`cart ${iconStyles.icon} ${styles["anchor-no-decoration"]}`}
              />
            </Link>
          </li>

          <li>
            <a href="#" target="_blank">
              <GitHubIcon className={`github ${iconStyles.icon}`} />
            </a>
          </li>
        </ul>
      </nav>
    </header>
  );
};

export default Header;
