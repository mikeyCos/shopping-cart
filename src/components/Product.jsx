import { useLocation } from "react-router-dom";
import { CartContext } from "./Cart";
import { useContext, useState } from "react";
import ProductCard from "./ProductCard";
import FormQuantity from "./FormQuantity";
import InputQuantity from "./InputQuantity";
import formatPrice from "../utilities/formatPrice";
import styles from "../styles/Product.module.css";

const Product = () => {
  const { state } = useLocation();
  const { product } = state;
  const { addToCart } = useContext(CartContext);
  const [quantity, setQuantity] = useState(1);
  const [quantityValidity, setQuantityValidity] = useState(true);

  const onSubmitHandler = () => {
    const isQuantityValid = validateQuantity(quantity, setQuantityValidity);
    isQuantityValid && addToCart(product, quantity);
  };

  const onChangeQuantityHandler = (newQuantity) => {
    setQuantity(newQuantity);
  };

  const incrementQuantityHandler = () => {
    setQuantity((prevQuantity) => {
      const newQuantity = +prevQuantity + 1;
      !quantityValidity && validateQuantity(newQuantity, setQuantityValidity);
      return newQuantity <= 999 ? newQuantity : prevQuantity;
    });
  };

  const decrementQuantityHandler = () => {
    setQuantity((prevQuantity) => {
      const newQuantity = prevQuantity - 1;
      return newQuantity <= 0 ? prevQuantity : newQuantity;
    });
  };

  const className =
    styles[state.previousLocation ? "product-modal" : "product"];

  return (
    <ProductCard className={className}>
      <ProductCard.Heading title={product.title} />
      <ProductCard.Picture src={product.image} alt="#" />
      <FormQuantity submitForm={() => onSubmitHandler()}>
        <p>{product.description}</p>
        <p>{formatPrice(product.price)}</p>
        <InputQuantity
          quantity={quantity}
          setQuantity={onChangeQuantityHandler}
          incrementHandler={incrementQuantityHandler}
          decrementHandler={decrementQuantityHandler}
        />
        <FormQuantity.SubmitButton text="Add to cart" />
      </FormQuantity>
    </ProductCard>
  );
};

const onInputValidate = (e) => {
  validateQuantity(e.target.value);
};

const validateQuantity = (quantity, callBack) => {
  const isQuantityValid = /^(?!^0)\d{1,3}$/.test(quantity);
  const input = document.querySelector(".quantity-input");
  const errorMessage = document.querySelector(".error-message");
  console.log("validateQuantity firing!");
  if (!isQuantityValid) {
    errorMessage.textContent =
      "Do not leave blank. Enter a number between 1 and 999.";
    input.addEventListener("input", onInputValidate);
    callBack && callBack(false);
  } else {
    errorMessage.textContent = "";
    input.removeEventListener("input", onInputValidate);
    callBack && callBack(true);
  }

  return isQuantityValid;
};

export default Product;
