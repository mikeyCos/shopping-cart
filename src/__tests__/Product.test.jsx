import { describe, expect, it, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import Product from "../components/Product";
import { product } from "./mocks";
import userEvent from "@testing-library/user-event";
import { CartProvider } from "../components/Cart";

describe("Product component", () => {
  it("Product component matches snapshot", () => {
    const initialEntries = [{ pathname: "/", state: { product } }];
    render(
      <MemoryRouter initialEntries={initialEntries}>
        <Product />
      </MemoryRouter>
    );

    const productCard = screen.getByRole("article");
    expect(productCard).toMatchSnapshot();
  });

  it("Initial quantity input is 1", () => {
    const initialEntries = [{ pathname: "/", state: { product } }];
    render(
      <MemoryRouter initialEntries={initialEntries}>
        <Product />
      </MemoryRouter>
    );

    const quantityInput = screen.getByRole("textbox");
    expect(quantityInput.value).toBe("1");
  });

  it("10 is typed into the quantity input", async () => {
    const user = userEvent.setup();
    const initialEntries = [{ pathname: "/", state: { product } }];
    render(
      <MemoryRouter initialEntries={initialEntries}>
        <Product />
      </MemoryRouter>
    );

    const quantityInput = screen.getByRole("textbox");

    await user.clear(quantityInput);
    await user.type(quantityInput, "10");

    expect(quantityInput.value).toBe("10");
  });

  it("Quantity input will have focus after clicking it", async () => {
    const user = userEvent.setup();
    const initialEntries = [{ pathname: "/", state: { product } }];
    render(
      <MemoryRouter initialEntries={initialEntries}>
        <Product />
      </MemoryRouter>
    );

    const quantityInput = screen.getByRole("textbox");

    await user.click(quantityInput);

    expect(quantityInput).toHaveFocus();
  });

  it("Clicking the incrementButton increases the quantity input's value by 1 ", async () => {
    const user = userEvent.setup();
    const initialEntries = [{ pathname: "/", state: { product } }];
    render(
      <CartProvider>
        <MemoryRouter initialEntries={initialEntries}>
          <Product />
        </MemoryRouter>
      </CartProvider>
    );

    const quantityInput = screen.getByRole("textbox");
    const incrementButton = screen.getByLabelText("increment quantity");

    await user.click(incrementButton);
    expect(quantityInput.value).toBe("2");
  });

  it("The up arrow increases the quantity input's value by 1", async () => {
    const user = userEvent.setup();
    const initialEntries = [{ pathname: "/", state: { product } }];
    render(
      <MemoryRouter initialEntries={initialEntries}>
        <Product />
      </MemoryRouter>
    );

    const quantityInput = screen.getByRole("textbox");

    await user.click(quantityInput);
    await user.keyboard("[ArrowUp]");
    expect(quantityInput.value).toBe("2");
  });
});
