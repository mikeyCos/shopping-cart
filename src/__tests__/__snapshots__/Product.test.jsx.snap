// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Product component > Product component matches snapshot 1`] = `
<article
  class="_product_b0e857"
>
  <h4>
    Opna Women's Short Sleeve Moisture
  </h4>
  <picture>
    <img
      alt="#"
      class="_product-img_c43253"
      loading="lazy"
      src="https://fakestoreapi.com/img/51eg55uWmdL._AC_UX679_.jpg"
    />
  </picture>
  <form
    novalidate=""
  >
    <p>
      100% Polyester, Machine wash, 100% cationic polyester interlock, Machine Wash & Pre Shrunk for a Great Fit, Lightweight, roomy and highly breathable with moisture wicking fabric which helps to keep moisture away, Soft Lightweight Fabric with comfortable V-neck collar and a slimmer fit, delivers a sleek, more feminine silhouette and Added Comfort
    </p>
    <p>
      $7.95
    </p>
    <div>
      <label
        for="quantity"
      />
      <div
        class="_input-btn-container_e589d5"
      >
        <button
          class="_input-btn-left_e589d5"
          type="button"
        >
          -
        </button>
        <input
          autocomplete="off"
          class="quantity-input _input_e589d5"
          id="quantity"
          maxlength="3"
          minlength="1"
          name="quantity"
          type="text"
          value="1"
        />
        <button
          class="_input-btn-right_e589d5"
          type="button"
        >
          +
        </button>
      </div>
      <p
        class="error-message"
      />
    </div>
    <button
      type="submit"
    >
      Add to cart
    </button>
  </form>
</article>
`;
